pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -235,20 +237,39 @@ private CompletableFuture<Reader<T>> readAllExistingMessages(Reader<T> reader) {
         AtomicLong messagesRead = new AtomicLong();
 
         CompletableFuture<Reader<T>> future = new CompletableFuture<>();
-        readAllExistingMessages(reader, future, startTime, messagesRead);
+        reader.getLastMessageIdsAsync().thenAccept(lastMessageIds -> {
+            Map<String, TopicMessageId> maxMessageIds = new ConcurrentHashMap<>();
+            lastMessageIds.forEach(topicMessageId -> {
+                maxMessageIds.put(topicMessageId.getOwnerTopic(), topicMessageId);
+            });
+            readAllExistingMessages(reader, future, startTime, messagesRead, maxMessageIds);
+        }).exceptionally(ex -> {
+            future.completeExceptionally(ex);
+            return null;
+        });
         return future;
     }
 
     private void readAllExistingMessages(Reader<T> reader, CompletableFuture<Reader<T>> future, long startTime,
-                                         AtomicLong messagesRead) {
+                                         AtomicLong messagesRead, Map<String, TopicMessageId> maxMessageIds) {
         reader.hasMessageAvailableAsync()
                 .thenAccept(hasMessage -> {
                    if (hasMessage) {
                        reader.readNextAsync()
                                .thenAccept(msg -> {
                                   messagesRead.incrementAndGet();
                                   handleMessage(msg);
-                                  readAllExistingMessages(reader, future, startTime, messagesRead);
+                                  // The message is read one by one in a single thread,
+                                  // so it's fine that uses a hashmap to store last message ID.
+                                  TopicMessageId maxMessageId = maxMessageIds.get(msg.getTopicName());
```

Comment:
I'm wondering how should we handle the case when `maxMessageId == null`? Currently I think it's impossible unless there is a bug somewhere else. In this case, I think we should print error logs.

Reply:
We need to remove it from the `maxMessageIds` once the partition has been read completely.
```java
        TopicMessageId maxMessageId = maxMessageIds.get(msg.getTopicName());
        if (maxMessageId != null && msg.getMessageId().compareTo(maxMessageId) >= 0) {
            maxMessageIds.remove(msg.getTopicName());
        }
```

Reply:
The received message could be as follows:
```
| Partition1 -> msg1 | Partition2 -> msg1 | Partition2 -> msg2 | Partition1 -> msg2|
```
We can not decide whether the updated completely by the max position. e.g. `Partition2-msg2`

Reply:
You didn't answer my question directly.

> how should we handle the case when maxMessageId == null?

I mean, should we treat `maxMessageId == null` as an unexpected case? Or is it okay to ignore this case

Reply:
> We need to remove it from the `maxMessageIds` once the partition has been read completely.
> ```java
>         TopicMessageId maxMessageId = maxMessageIds.get(msg.getTopicName());
>         if (maxMessageId != null && msg.getMessageId().compareTo(maxMessageId) >= 0) {
>             maxMessageIds.remove(msg.getTopicName());
>         }
> ```


@BewareMyPower I mean this is not an unexpected case.

Reply:
> We need to remove it from the `maxMessageIds` once the partition has been read completely.
> ```java
>         TopicMessageId maxMessageId = maxMessageIds.get(msg.getTopicName());
>         if (maxMessageId != null && msg.getMessageId().compareTo(maxMessageId) >= 0) {
>             maxMessageIds.remove(msg.getTopicName());
>         }
> ```


@BewareMyPower I mean this is not an unexpected case.

Reply:
> We need to remove it from the `maxMessageIds` once the partition has been read completely.
> ```java
>         TopicMessageId maxMessageId = maxMessageIds.get(msg.getTopicName());
>         if (maxMessageId != null && msg.getMessageId().compareTo(maxMessageId) >= 0) {
>             maxMessageIds.remove(msg.getTopicName());
>         }
> ```


@BewareMyPower I mean this is not an unexpected case.